consensus/src/creation/creator.rs

```diff
@@ -2,87 +2,134 @@ use crate::{
+    candidates: NodeMap<H::Hash>,
+    n_candidates: NodeCount,
+}
+
+impl<H: Hasher> UnitsCollector<H> {
+    pub fn new(n_members: NodeCount) -> Self {
+        Self {
+            candidates: NodeMap::with_size(n_members),
+            n_candidates: NodeCount(0),
+        }
+    }
+
+    pub fn add_unit(&mut self, unit: &Unit<H>) {
+        let node_ix = unit.creator();
+        let hash = unit.hash();
+
+        if self.candidates.get(node_ix).is_none() {
+            self.candidates.insert(node_ix, hash);
+            self.n_candidates += NodeCount(1);
+        }
+    }
+
+    pub fn prospective_parents(
+        &self,
+        node_ix: NodeIndex,
+    ) -> Result<&NodeMap<H::Hash>, ConstraintError> {
+        let threshold = (self.candidates.size() * 2) / 3 + NodeCount(1);
+
+        if self.n_candidates < threshold {
+            return Err(ConstraintError::NotEnoughParents);
+        }
+        if self.candidates.get(node_ix).is_none() {
+            return Err(ConstraintError::MissingOwnParent);
+        }
+        Ok(&self.candidates)
+    }
+}
+
+fn create_unit<H: Hasher>(
+    node_ix: NodeIndex,
+    parents: NodeMap<H::Hash>,
+    round: Round,
+) -> (PreUnit<H>, Vec<H::Hash>) {
+    let control_hash = ControlHash::new(&parents);
+    let parent_hashes = parents.into_values().collect();
+
+    let new_preunit = PreUnit::new(node_ix, round, control_hash);
+    (new_preunit, parent_hashes)
+}
 
 pub struct Creator<H: Hasher> {
-    node_id: NodeIndex,
+    round_collectors: Vec<UnitsCollector<H>>,
+    node_ix: NodeIndex,
     n_members: NodeCount,
-    candidates_by_round: Vec<NodeMap<H::Hash>>,
-    n_candidates_by_round: Vec<NodeCount>, // len of this - 1 is the highest round number of all known units
 }
 
 impl<H: Hasher> Creator<H> {
-    pub fn new(node_id: NodeIndex, n_members: NodeCount) -> Self {
+    pub fn new(node_ix: NodeIndex, n_members: NodeCount) -> Self {
         Creator {
-            node_id,
+            node_ix,
             n_members,
-            candidates_by_round: vec![NodeMap::with_size(n_members)],
-            n_candidates_by_round: vec![NodeCount(0)],
+            round_collectors: vec![UnitsCollector::new(n_members)],
         }
     }
 
     pub fn current_round(&self) -> Round {
-        (self.n_candidates_by_round.len() - 1) as Round
+        (self.round_collectors.len() - 1) as Round
     }
 
-    // initializes the vectors corresponding to the given round (and all between if not there)
-    fn init_round(&mut self, round: Round) {
-        if round > self.current_round() {
-            let new_size = (round + 1).into();
-            self.candidates_by_round
-                .resize(new_size, NodeMap::with_size(self.n_members));
-            self.n_candidates_by_round.resize(new_size, NodeCount(0));
-        }
+    // gets or initializes unit collectors for a given round (and all between if not there)
+    fn get_collector_for_round(&mut self, round: Round) -> &mut UnitsCollector<H> {
```

Comment:
How about naming it `get_or_initialize_collector_for_round`, calling it only in `add_unit` and using explicitly `current_round` in it, instead of `...len() -1` in that condition?

Reply:
Yes, that sounds good.